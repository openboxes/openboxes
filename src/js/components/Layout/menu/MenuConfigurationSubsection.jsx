import React from 'react';

import PropTypes from 'prop-types';

const MenuConfigurationSubsection = ({ subsection }) => (
  <div className="padding-8">
    <span className="subsection-section-title">{subsection.label && subsection.label}</span>
    {subsection.menuItems && subsection.menuItems.map((menuItem) => (
      <a className="dropdown-item" key={`${menuItem.label}-menuItem`} href={menuItem.href} target={menuItem.target}>
        {menuItem.label}
      </a>
    ))}
  </div>
);

export default MenuConfigurationSubsection;

const menuItemPropType = PropTypes.shape({
  label: PropTypes.string,
  href: PropTypes.string,
});

MenuConfigurationSubsection.propTypes = {
  subsection: PropTypes.shape({
    label: PropTypes.string,
    menuItems: PropTypes.arrayOf(menuItemPropType),
  }).isRequired,
};
