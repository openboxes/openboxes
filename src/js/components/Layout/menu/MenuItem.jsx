import React from 'react';

import _ from 'lodash';
import PropTypes from 'prop-types';
import { RiArrowDropDownLine } from 'react-icons/ri';

const DropdownMenuItem = ({ section, active }) => (
  <li className={`nav-item dropdown d-none d-md-flex justify-content-center align-items-center ${active && 'active-section'}`} data-testid="dropdownMenuItem">
    <a role="menuitem" className="nav-link dropdown-toggle" href="#" id="navbarDropdown" aria-haspopup="true" aria-expanded="false">
      {section.label}
    </a>
    <div className="dropdown-menu dropdown-menu-wrapper" aria-labelledby="navbarDropdown">
      <div className="dropdown-menu-content padding-8">
        {_.map(section.menuItems, (menuItem, menuItemKey) => (
          <a className="dropdown-item" key={menuItemKey} href={menuItem.href} target={menuItem.target}>
            {menuItem.label}
          </a>
        ))}
      </div>
    </div>
  </li>
);

const CollapseMenuItem = ({ section, active }) => {
  const id = `collapse-${section?.label?.replaceAll(' ', '-')}`;
  return (
    <li className="collapse-nav-item nav-item justify-content-center align-items-center d-flex d-md-none" data-testid="collapseMenuItem">
      <a
        className={`nav-link d-flex justify-content-between align-items-center w-100 ${active && 'active-section'}`}
        data-toggle="collapse"
        href={`#${id}`}
        role="button"
        aria-expanded="false"
        aria-controls={id}
      >
        {section.label}
        <RiArrowDropDownLine className="collapse-arrow-icon" />
      </a>
      <div className="collapse w-100" id={id}>
        <div className="d-flex flex-row flex-wrap">
          {_.map(section.menuItems, (menuItem, menuItemKey) => (
            <a className="subsection-section-item" key={menuItemKey} href={menuItem.href} target={menuItem.target}>
              {menuItem.label}
            </a>
          ))}
        </div>
      </div>
    </li>
  );
};

const MenuItem = (props) => (
  <>
    <DropdownMenuItem {...props} />
    <CollapseMenuItem {...props} />
  </>
);

export default MenuItem;

const menuItemPropType = PropTypes.shape({
  label: PropTypes.string,
  href: PropTypes.string,
});

MenuItem.propTypes = {
  section: PropTypes.shape({
    label: PropTypes.string,
    menuItems: PropTypes.arrayOf(menuItemPropType),
  }).isRequired,
  active: PropTypes.bool,
};

DropdownMenuItem.propTypes = { ...MenuItem.propTypes };
DropdownMenuItem.defaultProps = { ...MenuItem.defaultProps };

CollapseMenuItem.propTypes = { ...MenuItem.propTypes };
CollapseMenuItem.defaultProps = { ...MenuItem.defaultProps };

MenuItem.defaultProps = {
  active: false,
};
