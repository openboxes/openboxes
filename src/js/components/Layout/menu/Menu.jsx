import React, { useMemo } from 'react';

import _ from 'lodash';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { useParams, withRouter } from 'react-router-dom';

import MenuItem from 'components/Layout/menu/MenuItem';
import MenuSection from 'components/Layout/menu/MenuSection';
import MenuSubsection from 'components/Layout/menu/MenuSubsection';
import { checkActiveSection, getAllMenuUrls } from 'utils/menu-utils';

const Menu = ({ menuConfig, location, menuSectionsUrlParts }) => {
  const params = useParams();

  const allMenuUrls = useMemo(() => getAllMenuUrls(menuConfig), [menuConfig]);
  const activeSection = useMemo(() =>
    checkActiveSection({
      menuUrls: allMenuUrls,
      path: location,
      params,
      menuSectionsUrlParts,
    }), [allMenuUrls, location]);

  return (
    <div className="menu-wrapper" id="navbarSupportedContent">
      <ul className="d-flex align-items-center navbar-nav mr-auto flex-wrap">
        { _.chain(menuConfig)
          .filter((section) => section.id !== 'configuration')
          .map((section) => {
            if (section.href) {
              return (
                <MenuSection
                  section={section}
                  key={`${section.label}-menu-section`}
                  active={activeSection === section.id}
                />
              );
            }
            if (section.subsections) {
              return (
                <MenuSubsection
                  section={section}
                  key={`${section.label}-menu-subsection`}
                  active={activeSection === section.id}
                />
              );
            }
            if (section.menuItems) {
              return (
                <MenuItem
                  section={section}
                  key={`${section.label}-menuItem`}
                  active={activeSection === section.id}
                />
              );
            }
            return null;
          })
          .value()}
      </ul>
    </div>
  );
};

const mapStateToProps = (state) => ({
  menuConfig: state.session.menuConfig,
  menuSectionsUrlParts: state.session.menuSectionsUrlParts,
});

export default withRouter(connect(mapStateToProps)(Menu));

const menuItemPropType = PropTypes.shape({
  label: PropTypes.string,
  href: PropTypes.string,
});

const subsectionPropTypes = PropTypes.shape({
  label: PropTypes.string,
  menuItems: PropTypes.arrayOf(menuItemPropType),
});

const sectionPropTypes = PropTypes.shape({
  label: PropTypes.string,
  href: PropTypes.string,
  subsections: PropTypes.arrayOf(subsectionPropTypes),
  menuItems: PropTypes.arrayOf(menuItemPropType),
});

Menu.propTypes = {
  location: PropTypes.shape({
    pathname: PropTypes.string,
    search: PropTypes.string,
  }).isRequired,
  menuSectionsUrlParts: PropTypes.shape({
    inventory: PropTypes.arrayOf(PropTypes.string),
    products: PropTypes.arrayOf(PropTypes.string),
    purchasing: PropTypes.arrayOf(PropTypes.string),
    invoicing: PropTypes.arrayOf(PropTypes.string),
    inbound: PropTypes.arrayOf(PropTypes.string),
    outbound: PropTypes.arrayOf(PropTypes.string),
    requisitionTemplate: PropTypes.arrayOf(PropTypes.string),
    configuration: PropTypes.arrayOf(PropTypes.string),
    injectedDirectly: PropTypes.arrayOf(PropTypes.string),
  }).isRequired,
  menuConfig: PropTypes.arrayOf(sectionPropTypes).isRequired,
};
