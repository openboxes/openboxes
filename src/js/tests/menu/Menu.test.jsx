import React from 'react';

import { render, screen } from '@testing-library/react';
import { BrowserRouter as Router } from 'react-router-dom';
import renderer from 'react-test-renderer';

import Menu from 'components/Layout/menu/Menu';
import MenuConfigurationSubsection from 'components/Layout/menu/MenuConfigurationSubsection';
import MenuItem from 'components/Layout/menu/MenuItem';

import store from '../../store';

let renderedMenu;
let renderedMenuConfigurationSubsection;
let renderedMenuItem;
let subsection;

describe('menu component', () => {
  beforeEach(() => {
    const menuConfig = [{
      id: 'configuration',
      label: 'Configuration',
      href: '/openboxes/configuration/index',
    }];
    renderedMenu = renderer.create(
      <Router>
        <Menu
          menuConfig={menuConfig}
          location={[]}
          store={store}
        />
      </Router>,
    );
  });

  it('should match snapshot', () => {
    expect(renderedMenu.toJSON())
      .toMatchSnapshot();
  });
});

describe('menuConfigurationSubsection component', () => {
  beforeEach(() => {
    subsection = {
      label: 'Administration',
      menuItems: [{
        label: 'Settings',
        href: '/openboxes/admin/showSettings',
      }],
    };
    renderedMenuConfigurationSubsection = renderer
      .create(<Router><MenuConfigurationSubsection subsection={subsection} /></Router>);
  });

  it('should map menu items correctly', () => {
    render(<Router><MenuConfigurationSubsection subsection={subsection} /></Router>);
    expect(screen.findByText('Settings'))
      .toBeTruthy();
  });

  it('should match snapshot', () => {
    expect(renderedMenuConfigurationSubsection.toJSON())
      .toMatchSnapshot();
  });
});

describe('menuItem component', () => {
  beforeEach(() => {
    renderedMenuItem = renderer
      .create(<Router><MenuItem section={{}} active={false} /></Router>);
  });

  it('should match snapshot', () => {
    expect(renderedMenuItem.toJSON())
      .toMatchSnapshot();
  });

  it('should render collapseMenuItem and dropdownMenuItem correctly', () => {
    expect(renderedMenuItem.root.findByProps({ 'data-testid': 'collapseMenuItem' }))
      .toBeTruthy();
    expect(renderedMenuItem.root.findByProps({ 'data-testid': 'dropdownMenuItem' }))
      .toBeTruthy();
  });
});
