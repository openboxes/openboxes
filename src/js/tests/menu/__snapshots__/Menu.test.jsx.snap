// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`menu component should match snapshot 1`] = `
<div
  className="menu-wrapper"
  id="navbarSupportedContent"
>
  <ul
    className="d-flex align-items-center navbar-nav mr-auto flex-wrap"
  />
</div>
`;

exports[`menuConfigurationSubsection component should match snapshot 1`] = `
<div
  className="padding-8"
>
  <span
    className="subsection-section-title"
  >
    Administration
  </span>
  <a
    className="dropdown-item"
    href="/openboxes/admin/showSettings"
  >
    Settings
  </a>
</div>
`;

exports[`menuItem component should match snapshot 1`] = `
[
  <li
    className="nav-item dropdown d-none d-md-flex justify-content-center align-items-center false"
    data-testid="dropdownMenuItem"
  >
    <a
      aria-expanded="false"
      aria-haspopup="true"
      className="nav-link dropdown-toggle"
      href="#"
      id="navbarDropdown"
      role="menuitem"
    />
    <div
      aria-labelledby="navbarDropdown"
      className="dropdown-menu dropdown-menu-wrapper"
    >
      <div
        className="dropdown-menu-content padding-8"
      />
    </div>
  </li>,
  <li
    className="collapse-nav-item nav-item justify-content-center align-items-center d-flex d-md-none"
    data-testid="collapseMenuItem"
  >
    <a
      aria-controls="collapse-undefined"
      aria-expanded="false"
      className="nav-link d-flex justify-content-between align-items-center w-100 false"
      data-toggle="collapse"
      href="#collapse-undefined"
      role="button"
    >
      <svg
        className="collapse-arrow-icon"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          {
            "color": undefined,
          }
        }
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <g>
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
          <path
            d="M12 15l-4.243-4.243 1.415-1.414L12 12.172l2.828-2.829 1.415 1.414z"
          />
        </g>
      </svg>
    </a>
    <div
      className="collapse w-100"
      id="collapse-undefined"
    >
      <div
        className="d-flex flex-row flex-wrap"
      />
    </div>
  </li>,
]
`;
